### PR TITLE
Add ServerName to Vault Agent template config

### LINKED ISSUE #9183: Vault agent's template do not respect tls_server_name
**Describe the bug**
We have a vault server at one ip, which has a cert for a name and a second machine running agent, but connects to the vault server by ip and uses tls_server_name. This works fine for approle auth, but fails for a template.

**To Reproduce**
1. Set up vault server under $IP, with a valid cert for $NAME
2. Setup a approle auth and get login data
3. On another machine start agent with config see below
3. Template is not expanded, log shows error because of no valid cert for $IP

**Expected behavior**
tls_server_name should be also respected for templating

**Environment:**
* Vault Server Version (retrieve with `vault status`): 1.4.2
* Vault CLI Version (retrieve with `vault version`): 1.4.2
* Server Operating System/Architecture: Debian 10 / amd64

Vault agent configuration file(s):

```hcl
pid_file = "/home/vault/pidfile"

vault {
  address = "https://$IP:8200"
  tls_server_name = "$NAME"
}

auto_auth {
  method "approle" {
    config = {
      role_id_file_path = "/etc/vault/role-id"
      secret_id_file_path = "/etc/vault/secret-id"

      remove_secret_id_file_after_reading = false
    }
  }

  sink "file" {
    config = {
      path = "/home/vault/token"
    }
  }
}

cache {
  use_auto_auth_token = true
}

listener "unix" {
  address = "/home/vault/vault.socket"
  tls_disable = true
}

template {
  source = "XXX"
  destination = "YYY"
}
```

XXX template: 
```
{{ with secret "transit/keys/ZZZ" }}{{ .Data.data.keys }}{{end}}
```

If there is anything missing, please just ask.

### PR DESCRIPTION
**Pull request by pbar1**
Reopening the closed #9401 here:

> When preparing the consul-template Vault config, the provided
> TLSServerName (in the Vault agent config as tls_server_name) was not
> being passed along to the consul-template Vault config.
> 
> Using the VAULT_TLS_SERVER_NAME environment variable was working since
> the consul-template config building does consider it.
> 
> To fix that, pass it along when setting up the consul-template
> config. Add a test which verifies basic TLS connectivity to Vault,
> including passing along the server name.
> 
> Fixes #9183

### CODE REVIEW

**Comment by elsesiy** [comment]
@ncabatoff @calvn Any chance we can get this into `1.7.2`? We're using the agent injector and the SNI info is not propagated due to this issue. Thanks!

**Review by jasonodonnell — APPROVED** [review]
I tested this and it worked as expected, however, I saw weird behavior where auto-auth would hang indefinitely if a bogus SNI was set. I'm looking deeper into this but don't think this is a blocker on this specific PR.

**Comment by jasonodonnell** [comment]
Holding off on merging as we look into a possible bug uncovered by this.

**Review by calvn — APPROVED** [review]

**Comment by jasonodonnell** [comment]
@pbar1 We're ready to merge this but a [CHANGELOG file for this change is needed](https://github.com/hashicorp/vault/blob/master/CONTRIBUTING.md#changelog-entries). Once added we'll merge and backport this to the 1.7.x branch for the 1.7.2 release.

Thanks!

**Comment by pbar1** [comment]
Hey, I've been out for a few days. Sure - I'll add that changelog message.

File: changelog/11288.txt
Comment (jasonodonnell) [inline changelog/11288.txt:4]:
```suggestion
```

**Pull request merged**
